Add manifests for the web app

kimwnasptd opened the pull request:
**What this PR does / why we need it**:

Provides the manifests to install the KFServing UI #1327 . The manifests are integrated into the `/contrib` folder to go along all the manifests of the repo.

The use will either be able to install the app as standalone [ via the `default` kustomization ] or as part of Kubeflow via the `kubeflow` overlay. Although, if someone would like to make it work alongside Kubeflow they would have to also extend the CentralDashboard's [configmap](https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/config/centraldashboard-config.yaml) to add an entry for this app.

**Which issue(s) this PR fixes**:
Fixes #1505 

**Special notes for your reviewer**:
This PR makes the following changes and assumptions:
1. Uses the `models-web-app` name for the web app
2. Depending on the app's mode, it either uses:
    * the `knative-ingress-gateway.knative-serving` ingress gateway in the standalone case
    * the `kubeflow-gateway.kubeflow` gateway in case it is deployed as part of Kubeflow
3. Adds a sidecar and [AuthorizationPolicy](https://istio.io/latest/docs/reference/config/security/authorization-policy/) objects to control the traffic and who can query the app
4. Removes the `commonLabels` for `kustomize.component` and `app.kubernetes.io/component`. Since the app is a distinct component, under the `app: kfserving`, it will need to have it's own label/selectors
5. Extends the aggregated cluster permissions of the `kubeflow-{admin,editor,viewer}` roles to include Knative objects. In Kubeflow mode the UI exposes information about the owned Knative objects, so it will be checking if the user has the necessary permissions on these objects.

As I also mentioned in #1511 I don't have any hard preference on the app's name, so we could discuss if you think we could use a more fitting naming scheme.

Lastly if you would be interested in having a sidecar in the app in the standalone mode as well then I can add it.

kimwnasptd commented:
cc @yuzisun 
/assign @pvaneck

pvaneck commented on config/web-app/istio.yaml:
```diff
@@ -0,0 +1,20 @@
+  name: models-web-app
+spec:
+  gateways:
+  - knative-ingress-gateway.knative-serving
```
Can this be a variable `$(ingressGateway)` that uses the  `config/default/params.env` setting? With this, you wont need the VS gateway patch in the kubeflow overlay.  

Looking at `config/default/params.env` now, in order for you to use it, the following change ro params.env will need to be made: 

```
ingressGateway=knative-ingress-gateway.knative-serving
```

Then the `config/configmap/inferenceservice.yaml` will need the ingress section changed to:
`"ingressGateway" : "$(ingressGateway)"` 

Similarly, the quotes and comma will have to be removed from the kubeflow overlay params.env.

Probably should have been this way in the first place.

kimwnasptd replied:
Yes I would also prefer to use the `$(ingressGateway)` var, as it would simplify the manifests and we won't need to maintain an extra patch. Also thanks for the instructions!

I'll add a commit with the proposed refactor. I'll also give a heads up here if something comes up that we didn't catch with our first understanding.

kimwnasptd replied:
Pushed a commit with the described changes. Building both the `default` and the `kubeflow` artifacts results in using the correct gateway both in the VirtualSercice and in the ConfigMap.

I also removed the extra patch I had introduced to edit this field

pvaneck commented on config/web-app/istio.yaml:
```diff
@@ -0,0 +1,20 @@
+  http:
+  - match:
+    - uri:
+        prefix: /models/
```
Not a big deal, but I noticed when deploying standalone when visiting the app, if I didn't have the trailing slash, I'd get a 404. Is it okay to remove the trailing slash in the prefix?

kimwnasptd replied:
This makes sense, but it will be a little bit more involved than only changing the prefix in the VirtualService.

The [common code](https://github.com/kubeflow/kubeflow/blob/master/components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py#L20) sets the `<base href="${PREFIX}/">` to the `index.html` of all the web apps right now, and that prefix always has a backslash at the end.

Making the apps work without a trailing backslash makes sense for all of Kubeflow's web apps so I'll open an issue in `kubeflow/kubeflow` to track this. Until I fix this we will need to keep the backslash here as well.

Does this sound OK to you?

pvaneck replied:
Sure, that's fine. Can leave it as is.

kimwnasptd replied:
Created an issue to track this effort in kubeflow/kubeflow https://github.com/kubeflow/kubeflow/issues/5802

pvaneck reviewed (commented):
Hey, thanks for this! Left some comments inline. I guess the tests will be failing until we have the image created. 

Didn't have a chance yet to test integrated with Kubeflow (soon once the new manifests and 1.3 install instructions are settled), so I only tested the standalone version.

kimwnasptd commented on config/web-app/rbac.yaml:
```diff
@@ -0,0 +1,68 @@
+subjects:
+- kind: ServiceAccount
+  name: models-web-app
+  namespace: kubeflow
```
ACK, on it

kimwnasptd replied:
In the end I completely removed the namespace here as well, instead of using `kfserving-system`. If I kept the namespace definition in the `subject` then kustomize wouldn't do the substitution and change it to `kubeflow`.

kimwnasptd replied:
Pushed the commit for this, but will wait for an ACK/NAK first before resolving the conversation

pvaneck replied:
Interesting how this didn't work as KFServing has other clusterrolebindings with subjects that include the `kfserving-system` namespace that the overlay does successfully overwrite.  Looks like the `namespace: kfserving-system` in the `web-app/kustomization.yaml` is causing this.

Mind removing the namespace from the kustomization.yaml and explicitly adding the `namespace: kfserving-system` to the non-cluster resources under web-app (Deployment, VirtualService, Service, ServiceAccount)? This follows more in line with other config files in KFServing, and you should then also be able to keep the namespace in the subject.

kimwnasptd replied:
@pvaneck you were right on the money! Indeed after:
1. removing the namespace from the `kustomization.yaml` of the web-app
2. explicitly adding the `namespace: kfserving-system` in all the non-cluster resources, and patches

as you suggested worked. Kustomize can now successfully detect and modify the name accordingly. All the namespaces in all the base yamls as well as the patches are pointing to `kfserving-system` and are changed to `kubeflow` when building the corresponding overlay.

Again thanks for your help!

kimwnasptd replied:
Pushed another commit with this latest change

kimwnasptd commented on config/web-app/rbac.yaml:
```diff
@@ -0,0 +1,68 @@
+kind: ServiceAccount
+metadata:
+  name: models-web-app
+  namespace: kubeflow
```
Makes sense, will do

kimwnasptd replied:
Added a commit

kimwnasptd commented on config/web-app/service.yaml:
```diff
@@ -0,0 +1,14 @@
+kind: Service
+metadata:
+  name: models-web-app
+  namespace: kubeflow
```
Yep, on it

kimwnasptd replied:
Added a commit for this so I'll resolve the conversation

kimwnasptd commented on config/web-app/istio.yaml:
```diff
@@ -0,0 +1,20 @@
+      uri: /
+    route:
+    - destination:
+        host: models-web-app.kubeflow.svc.cluster.local
```
Indeed, my last copy paste bit me here. Thanks for catching it!
Will push a commit to fix this

kimwnasptd replied:
Added a commit that makes the default value to be `models-web-app.kfserving-system.svc.cluster.local` and I used a `patchJson6902` patch to override this value in the kubeflow overlay

pvaneck commented:
/retest

pvaneck reviewed (commented):
Tried this out with the Kubeflow 1.3 manifests and was able to see/deploy inferenceservices in my user namespace from the dashboard. :)

Metrics tab still shows with a bunch of "Could not connect to Grafana." since I do not have Grafana set up, so I'm not sure if the check to hide the tab or not is working as intended. However, that is a separate issue, and I think the manifests here look fine. Just need to get the CI to pass.

pvaneck commented:
/retest

yuzisun commented:
/retest

kimwnasptd commented:
> Tried this out with the Kubeflow 1.3 manifests and was able to see/deploy inferenceservices in my user namespace from the dashboard. :)

Glad to hear!

> Metrics tab still shows with a bunch of "Could not connect to Grafana." since I do not have Grafana set up, so I'm not sure if the check to hide the tab or not is working as intended. However, that is a separate issue, and I think the manifests here look fine. Just need to get the CI to pass.

I'll look at it again with the latest manifests for 1.3 and open an issue accordingly.
I'm also working on some basic manifests for exposing a minimal Grafana instance and integrate with KFServing's monitoring solution. I'll also open an issue for that effort as well.

kimwnasptd commented:
Let me also rebase the PR on top of the latest master, in case it helps with the e2e tests

kimwnasptd commented:
I can see the following error in the `run-e2e-tests(0)` task of the Argo Workflow graph:
```
>           raise RuntimeError("Timeout to start the InferenceService {}. \
                               The InferenceService is as following: {}".format(name, current_isvc))
E           RuntimeError: Timeout to start the InferenceService isvc-explainer-tabular.                                The InferenceServi
ce is as following: {'apiVersion': 'serving.kubeflow.org/v1beta1', 'kind': 'InferenceService', 'metadata': ...
...'name': 'isvc-explainer-tabular', 'namespace': 'kfserving-ci-e2e-test', ...
```

The test also took about an hour to finish. It seems that changing the manifests results in a lot of components to be tested at once, which is also expected.

/retest

kimwnasptd commented:
Hmm I see the exact same error for the `isvc-explainer-tabular` InferenceService. Looking at the status I also see:
```
{'lastTransitionTime': '2021-04-14T15:39:34Z', 'reason': 'Explainer ingress not created', 'status': 'False', 'type': 'Ready'}
```

Is this expected? Any ideas to make the tests pass?

/retest

kimwnasptd commented:
Τhe tests passed. I'm not going to touch anything :)

pvaneck commented:
/lgtm

yuzisun commented on config/overlays/kubeflow/patches/web-app-sidecar.yaml:
```diff
@@ -0,0 +1,10 @@
+kind: Deployment
+metadata:
+  name: models-web-app
+  namespace: kfserving-system
```
I think the namespace is[ kustomized](https://github.com/kubeflow/kfserving/blob/master/config/overlays/kubeflow/kustomization.yaml#L5) so we do not need to specify here

kimwnasptd replied:
Removing the namespace from that line would result in
```
Error: no matches for OriginalId apps_v1_Deployment|~X|models-web-app; no matches for CurrentId apps_v1_Deployment|~X|models-web-app; failed to find unique target for patch apps_v1_Deployment|models-web-app
```
since I specifically set the `namespace: kfserving-system` to all the [base/default yamls](https://github.com/kubeflow/kfserving/tree/df64bbc5be6c07f418e0d1d7ce34c16c8d89887e/config/web-app) of the app.

To achieve what you describe I would have to remove the explicit mention of `namespace: kfserving-system` from all the web app's base/default yamls, introduced in this PR.

We slightly went over this with @pvaneck as well https://github.com/kubeflow/kfserving/pull/1512#discussion_r604501150 and in the end decided to stick with using the `namespace: kfserving-system` in the base yamls since this is what the other components are doing as well [ [controller](https://github.com/kubeflow/kfserving/blob/master/config/manager/manager.yaml#L5), [cert-manager](https://github.com/kubeflow/kfserving/blob/master/config/certmanager/certificate.yaml#L7), [configmap](https://github.com/kubeflow/kfserving/blob/master/config/configmap/inferenceservice.yaml#L5), [current sts patch](https://github.com/kubeflow/kfserving/blob/7fc4bbbbd598762536bc9b2965786b7cb310746b/config/overlays/kubeflow/statefulset-patch.yaml#L5) etc ].

yuzisun commented on config/web-app/istio.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: networking.istio.io/v1alpha3
```
```suggestion
apiVersion: networking.istio.io/v1beta1
```

kimwnasptd replied:
Good catch, thanks!

kimwnasptd replied:
I also added an extra commit since we will also need to use `v1beta1` in the patch selector, inside the `overlays/kubeflow/kustomization.yaml` file, as well.

yuzisun reviewed (commented):
@kimwnasptd a few nits other looks good! I am going to cut 0.6 RC release, is there anything you want to get in before that?

kimwnasptd commented:
Thanks for taking a look and for the heads up @yuzisun!

> I am going to cut 0.6 RC release, is there anything you want to get in before that?

I would only like to include https://github.com/kubeflow/kfserving/pull/1541 which resolves the issue @pvaneck bumped into as well https://github.com/kubeflow/kfserving/pull/1512#pullrequestreview-635140387 #1540, where the `METRICS` tab is shown even if Grafana is not exposed. This should improve the UX of users that will try out the app.

I believe this should be OK for exposing this alpha feature to users and get their feedback for now. In the meantime I'm almost done with defining some e2e tests #1507, but this will require some more friction with integrating with the CI so I don't think this will make it in time for the RC.

Would that be OK? Hope I'm not delaying the schedule.

pvaneck commented:
As I was testing the grafana tab PR with the rest of Kubeflow and listed pods in the kubeflow namespace, I initially thought the web-app didn't come up because I was looking for pods next to the `kfserving-controller-manager-0` pod. I am wondering if it would make more sense to rename instances of `models-web-app` to  `kfserving-models-web-app` or  `kfserving-web-app` in order to keep the kfserving resources next to each other when listing resources. This also make it more clear that a resource came from the KFServing component. What do you think? @kimwnasptd

kimwnasptd commented:
I don't have a strong opinion, since I was doing `k get <resource> -n kubeflow -l app=kfserving` to list objects owned by KFServing.

But it makes sense to also add a common `kfserving-` prefix to the web app's manifests to align with the other manifests and make it easy to the eye to locate all the KFServing components with a glance. Adding a commit and will also rebase to latest master, since there seems to be a conflict.

kimwnasptd commented:
@pvaneck I added a `kfserving-` prefix in front of the web apps manifests. Let me know if you'd like me to implement any more changes.

@yuzisun could you also type a  `@googlebot I consent.` comment since I rebased and used a suggested commit you proposed?

pvaneck commented:
@kimwnasptd Thanks! Validated again on IKS.

/lgtm
/retest

yuzisun commented:
@googlebot I consent

yuzisun commented:
/approve

yuzisun commented:
@googlebot I consent

kimwnasptd commented:
@yuzisun I believe you will also need to have a fullstop at the end of the message to register the command
`@googlebot I consent.`

Could you try that as well?

yuzisun commented:
@googlebot I consent.

kimwnasptd commented:
/retest

kimwnasptd commented:
/retest

kimwnasptd commented:
/retest

kimwnasptd commented:
/retest

yuzisun commented:
/lgtm
